### PR TITLE
update CLI Hacking Guide

### PR DESCRIPTION
**Pull request by juanvallejo**
Updates the CLI Hacking Guide with current conventions.

cc @soltysh

### CODE REVIEW

File: docs/cli_hacking_guide.adoc [hunk 1]
```diff
@@ -67,52 +73,68 @@ var (
...
-  options := &MineOptions{}
+func NewCmdMine(name, fullName string, f *clientcmd.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+  o := NewMineOptions(streams)
   cmd := &cobra.Command{
     Use:     fmt.Sprintf("%s [--latest]", name),
     Short:   "Run my command",
     Long:    mineLong,
     Example: fmt.Sprintf(mineExample, fullName),
     Run: func(cmd *cobra.Command, args []string) {
-      if err := options.Complete(f, cmd, args, out); err != nil { (6)
```
Comment (soltysh) [inline docs/cli_hacking_guide.adoc]: Change all that into:
```go
      kcmdutil.CheckErr(o.Complete(f, cmd, args))   (6)
      kcmdutil.CheckErr(o.Validate())   (7)
      kcmdutil.CheckErr(o.Run()    (8)
```

And don't pass `out` :stuck_out_tongue_closed_eyes:

File: docs/cli_hacking_guide.adoc [hunk 2]
```diff
@@ -257,62 +305,72 @@ var (
...
-        cmdutil.CheckErr(cmdutil.UsageErrorf(cmd, err.Error()))
-      }
+      cmdutil.CheckErr(o.Complete(f, cmd, args, out))
 
       // 8.
-      if err := options.RunDeploy(); err != nil {
-        // 9.
-        cmdutil.CheckErr(err)
-      }
+      cmdutil.CheckErr(o.Validate())
```
Comment (soltysh) [inline docs/cli_hacking_guide.adoc]: Use `kcmdutil` consistently across all the examples.

File: docs/cli_hacking_guide.adoc [hunk 3]
```diff
@@ -257,62 +305,72 @@ var (
...
-<9> Try to always use the functions in `k8s.io/kubernetes/pkg/kubectl/cmd/util` to check and handle errors. It is not expected that commands call `glog.Fatalf`, `os.Exit` or anything similar directly.
+<4> Create a "constructor" for the command options struct. Here you will instantiate command options, default any values, and set the IO streams for writing to the screen. 
+<5> This function creates the command. Notice it takes the parent command name as argument and also a `io.Writer` that will be used to print messages.
+<6> Command usage.
+<7> `Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string, out io.Writer) error` is used to populate any object or variable that will be required to run the command and is still missing at this point. For example, if the command will make use of an API client it can be created from the factory in this method. Can also be used to take argument values from the `args` slice and hold it in explicit variables in your struct, store the `io.Writer` that will be used later, etc.
```
Comment (soltysh) [inline docs/cli_hacking_guide.adoc]: Remove `out` from method declaration.

File: docs/cli_hacking_guide.adoc [hunk 4]
```diff
@@ -199,7 +221,26 @@ $ hack/update-generated-docs.sh
...
+  Run: func(cmd *cobra.Command, args []string) {
+    kcmdutil.CheckErr(o.Complete(f, cmd, args))
+    kcmdutil.CheckErr(o.Validate())
+    kcmdutil.CheckErr(o.Run())
+  },
+}
+----
+====
 
 === Tests
```
Comment (soltysh) [inline docs/cli_hacking_guide.adoc]: Remove the section about tests, at least for now. When we'll write it we'll add it :-)

**Comment by juanvallejo** [comment]
@soltysh thanks, review comments addressed

File: docs/cli_hacking_guide.adoc [hunk 5]
```diff
@@ -67,52 +73,62 @@ var (
...
-      if err := options.Complete(f, cmd, args, out); err != nil { (6)
-        cmdutil.CheckErr(err)
-      }
-      if err := options.Validate(args); err != nil { (7)
-        cmdutil.CheckErr(cmdutil.UsageErrorf(cmd, err.Error()))
-      }
-      if err := options.RunDeploy(); err != nil { (8)
-        cmdutil.CheckErr(err) (9)
-      }
+      cmdutil.CheckErr(o.Complete(f, cmd, args))
```
Comment (soltysh) [inline docs/cli_hacking_guide.adoc]: `kcmdutil` to be consistent with all the others.

**Comment by juanvallejo** [comment]
@soltysh thanks, review comment addressed

**Pull request merged**